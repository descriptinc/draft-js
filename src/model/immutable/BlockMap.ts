/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import {BlockNodeRecord} from './BlockNodeRecord';
import {flatMap, flatten} from '../descript/Iterables';

export type BlockMap = ReadonlyMap<string, BlockNodeRecord>;

// FIXME [mvp]: dealing with order of `updates`

export function mergeMapUpdates<T>(
  originalMap: ReadonlyMap<string, T>,
  updates: Record<string, T | null>,
): ReadonlyMap<string, T> {
  let didChange = false;
  const seenKeys = new Set<string>();
  const result: ReadonlyMap<string, T> = new Map(
    flatten([
      flatMap(originalMap, (entry): [string, T] | undefined => {
        const [key, existingBlock] = entry;
        seenKeys.add(key);
        const newBlock = updates[key];
        if (newBlock === null) {
          // remove this block from the map
          return undefined;
        }
        if (!newBlock || newBlock === existingBlock) {
          return entry;
        }
        didChange = true;
        return [key, newBlock];
      }),
      flatMap(Object.entries(updates), (entry): [string, T] | undefined => {
        const [key, block] = entry;
        if (!block || seenKeys.has(key)) {
          return undefined;
        }
        didChange = true;
        return [key, block];
      }),
    ]),
  );
  return didChange ? result : originalMap;
}
