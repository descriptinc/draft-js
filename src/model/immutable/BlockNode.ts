/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {DraftBlockType} from '../constants/DraftBlockType';
import {CharacterMetadata} from './CharacterMetadata';

export type BlockNodeKey = string;

export type BlockNode = Readonly<{
  characterList: readonly CharacterMetadata[];
  data: Readonly<Record<string, any>>;
  depth: number;
  key: BlockNodeKey;
  text: string;
  type: DraftBlockType;
}>;
