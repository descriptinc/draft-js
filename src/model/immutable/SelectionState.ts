/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

export type SelectionState = Readonly<{
  anchorKey: string;
  anchorOffset: number;
  focusKey: string;
  focusOffset: number;
  isBackward: boolean;
  hasFocus: boolean;
}>;

export function getStartOffset({
  isBackward,
  focusOffset,
  anchorOffset,
}: SelectionState): number {
  return isBackward ? focusOffset : anchorOffset;
}

export function getEndOffset({
  isBackward,
  anchorOffset,
  focusOffset,
}: SelectionState): number {
  return isBackward ? anchorOffset : focusOffset;
}

export function isCollapsed({
  anchorKey,
  focusKey,
  anchorOffset,
  focusOffset,
}: SelectionState): boolean {
  return anchorKey === focusKey && anchorOffset === focusOffset;
}

export function getStartKey({
  isBackward,
  focusKey,
  anchorKey,
}: SelectionState): string {
  return isBackward ? focusKey : anchorKey;
}

export function getEndKey({
  isBackward,
  focusKey,
  anchorKey,
}: SelectionState): string {
  return isBackward ? anchorKey : focusKey;
}

/**
 * Return whether the specified range overlaps with an edge of the
 * SelectionState.
 */
export function hasEdgeWithin(
  sel: SelectionState,
  blockKey: string,
  start: number,
  end: number,
): boolean {
  const {anchorKey, focusKey, anchorOffset, focusOffset} = sel;
  if (anchorKey === focusKey && anchorKey === blockKey) {
    const selectionStart = getStartOffset(sel);
    const selectionEnd = getEndOffset(sel);
    return (
      (start <= selectionStart && selectionStart <= end) || // selectionStart is between start and end, or
      (start <= selectionEnd && selectionEnd <= end) // selectionEnd is between start and end
    );
  }

  if (blockKey !== anchorKey && blockKey !== focusKey) {
    return false;
  }

  const offsetToCheck = blockKey === anchorKey ? anchorOffset : focusOffset;

  return start <= offsetToCheck && end >= offsetToCheck;
}

export function makeSelectionState({
  anchorKey,
  anchorOffset,
  focusKey,
  focusOffset,
  isBackward = false,
  hasFocus = false,
}: Partial<SelectionState> &
  Pick<
    SelectionState,
    'anchorKey' | 'anchorOffset' | 'focusKey' | 'focusOffset'
  >): SelectionState {
  return {
    anchorKey,
    anchorOffset,
    focusKey,
    focusOffset,
    isBackward,
    hasFocus,
  };
}

export function makeNullSelection(): SelectionState {
  return makeEmptySelection('');
}

export function makeEmptySelection(key: string): SelectionState {
  return makeSelectionState({
    anchorKey: key,
    anchorOffset: 0,
    focusKey: key,
    focusOffset: 0,
  });
}
