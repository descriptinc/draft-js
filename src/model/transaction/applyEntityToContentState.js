/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import type ContentState from 'ContentState';
import type SelectionState from 'SelectionState';

const applyEntityToContentBlock = require('applyEntityToContentBlock');

function applyEntityToContentState(
  contentState: ContentState,
  selectionState: SelectionState,
  entityKey: ?string,
): ContentState {
  const blockMap = contentState.getBlockMap();
  const startKey = selectionState.getStartKey();
  const startOffset = selectionState.getStartOffset();
  const endKey = selectionState.getEndKey();
  const endOffset = selectionState.getEndOffset();

  let newBlocks: ?Seq.Keyed<ContentBlock>;
  if (startKey !== endKey) {
    newBlocks = blockMap
      .toSeq()
      .skipUntil((_, k) => k === startKey)
      .takeUntil((_, k) => k === endKey)
      .map((block: ContentBlock, blockKey: string) => {
        const sliceStart = blockKey === startKey ? startOffset : 0;
        const sliceEnd = block.getLength();
        return applyEntityToContentBlock(
          block,
          sliceStart,
          sliceEnd,
          entityKey,
        );
      });
  }

  return contentState.merge({
    blockMap: blockMap.merge(newBlocks, {
      [endKey]: applyEntityToContentBlock(
        blockMap.get(endKey),
        endKey === startKey ? startOffset : 0,
        endOffset,
        entityKey,
      ),
    }),
    selectionBefore: selectionState,
    selectionAfter: selectionState,
  });
}

module.exports = applyEntityToContentState;
