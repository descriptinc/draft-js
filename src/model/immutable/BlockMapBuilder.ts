/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import {BlockNodeRecord} from './BlockNodeRecord';
import {BlockMap} from './BlockMap';
import {map} from '../descript/Iterables';

export function createFromArray(blocks: Iterable<BlockNodeRecord>): BlockMap {
  return new Map(
    map(blocks, (block): [string, BlockNodeRecord] => [block.key, block]),
  );
}
