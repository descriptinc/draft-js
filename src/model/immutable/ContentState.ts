/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {BlockMap} from './BlockMap';
import {
  SelectionState,
  makeEmptySelection,
  makeNullSelection,
} from './SelectionState';
import {first, last, rest, skipUntil, takeUntil} from '../descript/Iterables';
import {BlockNodeRecord} from './BlockNodeRecord';
import {DraftEntityType} from '../entity/DraftEntityType';
import {DraftEntityMutability} from '../entity/DraftEntityMutability';
import {createFromArray} from './BlockMapBuilder';
import {makeContentBlockNode} from './ContentBlockNode';
import {makeContentBlock} from './ContentBlock';
import {gkx} from '../../stubs/gkx';
import DraftEntity, {DraftEntityMapObject} from '../entity/DraftEntity';
import {DraftEntityInstance} from '../entity/DraftEntityInstance';
import sanitizeDraftText from '../encoding/sanitizeDraftText';

export type ContentState = Readonly<{
  blockMap: BlockMap;
  selectionBefore: SelectionState;
  selectionAfter: SelectionState;
}>;

export function makeContentState({
  blockMap,
  selectionAfter = makeEmptySelection(first(blockMap.keys())!),
  selectionBefore = makeEmptySelection(first(blockMap.keys())!),
}: Partial<ContentState> & Pick<ContentState, 'blockMap'>): ContentState {
  // FIXME [perf]: force type to be exact and don't create a new object?
  return {
    blockMap,
    selectionBefore,
    selectionAfter,
  };
}

export function getLastCreatedEntityKey(): string {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__getLastCreatedEntityKey();
}

export function getBlockForKey(
  {blockMap}: ContentState,
  key: string,
): BlockNodeRecord {
  const block = blockMap.get(key);
  if (!block) {
    throw new Error('No block found for key');
  }
  return block;
}

export function getFirstBlock({blockMap}: ContentState): BlockNodeRecord {
  const block = first(blockMap.values());
  if (!block) {
    throw new Error('Block map is empty');
  }
  return block;
}

export function getLastBlock({blockMap}: ContentState): BlockNodeRecord {
  const block = last(blockMap.values()); // FIXME [perf]: O(n)
  if (!block) {
    throw new Error('Block map is empty');
  }
  return block;
}

export function hasText(content: ContentState): boolean {
  const {blockMap} = content;
  if (blockMap.size > 1) {
    return true;
  }
  return (
    blockMap.size > 1 ||
    (blockMap.size > 0 &&
      escape(getFirstBlock(content).text).replace(/%u200B/g, '').length > 0)
  );
}

export function createEntity(
  type: DraftEntityType,
  mutability: DraftEntityMutability,
  data?: Object,
): string {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__create(type, mutability, data);
}

export function mergeEntityData(
  key: string,
  toMerge: Record<string, any>,
): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__mergeData(key, toMerge);
}

export function replaceEntityData(
  key: string,
  newData: Record<string, any>,
): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__replaceData(key, newData);
}

export function addEntity(instance: DraftEntityInstance): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__add(instance);
}

export function getEntity(key: string): DraftEntityInstance {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__get(key);
}

export function createFromBlockArray(
  blocks: BlockNodeRecord[] | {contentBlocks: BlockNodeRecord[]},
): ContentState {
  // TODO: remove this when we completely deprecate the old entity API
  const theBlocks = Array.isArray(blocks) ? blocks : blocks.contentBlocks;

  const blockMap = createFromArray(theBlocks);
  const firstBlock = first(blockMap.values());
  const selectionState = !firstBlock
    ? makeNullSelection()
    : makeEmptySelection(firstBlock.key);

  return makeContentState({
    blockMap,
    selectionBefore: selectionState,
    selectionAfter: selectionState,
  });
}

export function createFromText(
  text: string,
  delimiter: string | RegExp = /\r\n?|\n/g,
): ContentState {
  const strings = text.split(delimiter);
  const blocks = strings.map(block => {
    block = sanitizeDraftText(block);
    if (gkx('draft_tree_data_support')) {
      return makeContentBlockNode({
        text: block,
      });
    } else {
      return makeContentBlock({
        text: block,
      });
    }
  });
  return createFromBlockArray(blocks);
}

export function getBlockBefore(
  {blockMap}: ContentState,
  blockKey: string,
): BlockNodeRecord | undefined {
  // FIXME [perf]: cache
  const before = last(takeUntil(blockMap, ([k]) => k === blockKey));
  return before?.[1];
}

export function getBlockAfter(
  {blockMap}: ContentState,
  blockKey: string,
): BlockNodeRecord | undefined {
  // FIXME [perf]: cache
  const after = first(rest(skipUntil(blockMap, ([k]) => k === blockKey)));
  return after?.[1];
}

export function getEntityMap(_: ContentState): DraftEntityMapObject {
  return DraftEntity;
}
