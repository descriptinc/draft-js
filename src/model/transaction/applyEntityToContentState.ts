/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 */
import {ContentState} from '../immutable/ContentState';
import {
  getEndKey,
  getEndOffset,
  getStartKey,
  getStartOffset,
  SelectionState,
} from '../immutable/SelectionState';
import {flatten, map, skipUntil, takeUntil} from '../descript/Iterables';
import applyEntityToContentBlock from './applyEntityToContentBlock';
import {
  makeBlockMapIndex,
  BlockMapIndex,
  mergeMapUpdates,
  BlockMap,
} from '../immutable/BlockMap';
import {BlockNode} from '../immutable/BlockNode';

function updateNewBlocksWithEntity(
  blockMap: BlockMap,
  blockMapIndex: BlockMapIndex,
  newBlocks: Record<string, BlockNode>,
  selectionState: SelectionState,
  entityKey: string | null,
): void {
  const startKey = getStartKey(selectionState);
  const startOffset = getStartOffset(selectionState);
  const endKey = getEndKey(selectionState);
  const endOffset = getEndOffset(selectionState);

  let blockKey: string | null = startKey;
  while (blockKey) {
    const block = newBlocks[blockKey] || blockMap.get(blockKey);
    const sliceStart = blockKey === startKey ? startOffset : 0;
    const sliceEnd = blockKey === endKey ? endOffset : block.text.length;
    newBlocks[blockKey] = applyEntityToContentBlock(
      block,
      sliceStart,
      sliceEnd,
      entityKey,
    );
    if (blockKey === endKey) {
      return;
    }
    blockKey = blockMapIndex[blockKey];
  }
}

export default function applyEntityToContentState(
  contentState: ContentState,
  selectionState: SelectionState,
  entityKey: string | null,
): ContentState {
  const blockMap = contentState.blockMap;
  const startKey = getStartKey(selectionState);
  const startOffset = getStartOffset(selectionState);
  const endKey = getEndKey(selectionState);
  const endOffset = getEndOffset(selectionState);

  const iter = map(
    flatten<[string, BlockNode]>([
      takeUntil(
        skipUntil(blockMap, ([k]) => k === startKey),
        ([k]) => k === endKey,
      ),
      [[endKey, blockMap.get(endKey)!]],
    ]),
    ([blockKey, block]): [string, BlockNode] => {
      const sliceStart = blockKey === startKey ? startOffset : 0;
      const sliceEnd = blockKey === endKey ? endOffset : block.text.length;
      return [
        blockKey,
        applyEntityToContentBlock(block, sliceStart, sliceEnd, entityKey),
      ];
    },
  );

  const newBlocks: Record<string, BlockNode> = {};
  for (const [key, block] of iter) {
    newBlocks[key] = block;
  }

  return {
    ...contentState,
    blockMap: mergeMapUpdates(blockMap, newBlocks),
    selectionBefore: selectionState,
    selectionAfter: selectionState,
  };
}

export function applyEntitiesToContentState(
  contentState: ContentState,
  entities: Iterable<[SelectionState, string | null]>,
): ContentState {
  const blockMap = contentState.blockMap;
  const blockMapIndex = makeBlockMapIndex(blockMap);
  const newBlocks: Record<string, BlockNode> = {};
  let lastSelectionState: SelectionState | undefined;
  for (const [selectionState, entityKey] of entities) {
    updateNewBlocksWithEntity(
      blockMap,
      blockMapIndex,
      newBlocks,
      selectionState,
      entityKey,
    );
    lastSelectionState = selectionState;
  }

  return {
    ...contentState,
    blockMap: mergeMapUpdates(blockMap, newBlocks),
    ...(lastSelectionState
      ? {
          selectionBefore: lastSelectionState,
          selectionAfter: lastSelectionState,
        }
      : {}),
  };
}
