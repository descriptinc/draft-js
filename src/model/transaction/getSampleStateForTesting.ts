/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import {
  createWithContent,
  EditorState,
  forceSelection,
} from '../immutable/EditorState';
import {ContentState, makeContentState} from '../immutable/ContentState';
import {makeSelectionState, SelectionState} from '../immutable/SelectionState';
import {makeContentBlock} from '../immutable/ContentBlock';
import {repeat} from '../descript/Iterables';
import {makeCharacterMetadata} from '../immutable/CharacterMetadata';
import SampleInlineStyles from '../immutable/SampleDraftInlineStyle';
import {createFromArray} from '../immutable/BlockMapBuilder';
import DraftEntity from '../entity/DraftEntity';

const {BOLD, ITALIC} = SampleInlineStyles;
const ENTITY_KEY = '2';

const BLOCKS = [
  makeContentBlock({
    key: 'a',
    type: 'unstyled',
    text: 'Alpha',
  }),
  makeContentBlock({
    key: 'b',
    type: 'unordered-list-item',
    text: 'Bravo',
    characterList: Array.from(
      repeat(
        5,
        makeCharacterMetadata({
          style: BOLD,
          entity: ENTITY_KEY,
        }),
      ),
    ),
  }),
  makeContentBlock({
    key: 'c',
    type: 'code-block',
    text: 'Test',
  }),
  makeContentBlock({
    key: 'd',
    type: 'code-block',
    text: '',
    characterList: [],
  }),
  makeContentBlock({
    key: 'e',
    type: 'code-block',
    characterList: [],
  }),
  makeContentBlock({
    key: 'f',
    type: 'blockquote',
    text: 'Charlie',
    characterList: Array.from(
      repeat(
        7,
        makeCharacterMetadata({
          style: ITALIC,
          entity: null,
        }),
      ),
    ),
  }),
];

const selectionState = makeSelectionState({
  anchorKey: 'a',
  anchorOffset: 0,
  focusKey: 'a',
  focusOffset: 0,
  isBackward: false,
  hasFocus: true,
});

const blockMap = createFromArray(BLOCKS);
const contentState = makeContentState({
  blockMap,
  selectionBefore: selectionState,
  selectionAfter: selectionState,
});

DraftEntity.__create('IMAGE', 'IMMUTABLE');

let editorState = createWithContent(contentState);
editorState = forceSelection(editorState, selectionState);

const getSampleStateForTesting = (): {
  editorState: EditorState;
  contentState: ContentState;
  selectionState: SelectionState;
} => {
  return {editorState, contentState, selectionState};
};
export default getSampleStateForTesting;
