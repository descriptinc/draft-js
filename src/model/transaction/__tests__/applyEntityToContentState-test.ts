/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */
import getSampleStateForTesting from '../getSampleStateForTesting';
import {makeSelectionState} from '../../immutable/SelectionState';
import {getBlockAfter, getFirstBlock} from '../../immutable/ContentState';
import applyEntityToContentState from '../applyEntityToContentState';

const {contentState, selectionState} = getSampleStateForTesting();

const initialBlock = getFirstBlock(contentState);
const secondBlock = getBlockAfter(contentState, initialBlock.key);

const selectBlock = makeSelectionState({
  anchorKey: initialBlock.key,
  anchorOffset: 0,
  focusKey: initialBlock.key,
  focusOffset: initialBlock.text.length,
});

const selectAdjacentBlocks = makeSelectionState({
  anchorKey: initialBlock.key,
  anchorOffset: 0,
  focusKey: secondBlock?.key,
  focusOffset: secondBlock?.text.length,
});

const assertApplyEntityToContentState = (
  entityKey: string | null,
  selection = selectionState,
  content = contentState,
) => {
  expect(
    applyEntityToContentState(content, selection, entityKey).blockMap,
  ).toMatchSnapshot();
};

test('must apply entity key', () => {
  assertApplyEntityToContentState('x', selectBlock);
});

test('must apply null entity', () => {
  assertApplyEntityToContentState(null, selectBlock);
});

test('must apply entity key accross multiple blocks', () => {
  assertApplyEntityToContentState('x', selectAdjacentBlocks);
});

test('must apply null entity key accross multiple blocks', () => {
  assertApplyEntityToContentState(null, selectAdjacentBlocks);
});
