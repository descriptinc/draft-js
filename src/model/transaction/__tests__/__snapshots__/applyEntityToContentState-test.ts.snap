// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must apply entity key 1`] = `
Map {
  "a" => Object {
    "characterList": Array [
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "a",
    "text": "Alpha",
    "type": "unstyled",
  },
  "b" => Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;

exports[`must apply entity key accross multiple blocks 1`] = `
Map {
  "a" => Object {
    "characterList": Array [
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
      Object {
        "entity": "x",
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "a",
    "text": "Alpha",
    "type": "unstyled",
  },
  "b" => Object {
    "characterList": Array [
      Object {
        "entity": "x",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "x",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "x",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "x",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "x",
        "style": Set {
          "BOLD",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;

exports[`must apply null entity 1`] = `
Map {
  "a" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "a",
    "text": "Alpha",
    "type": "unstyled",
  },
  "b" => Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": "2",
        "style": Set {
          "BOLD",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;

exports[`must apply null entity key accross multiple blocks 1`] = `
Map {
  "a" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "a",
    "text": "Alpha",
    "type": "unstyled",
  },
  "b" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "BOLD",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "BOLD",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
      Object {
        "entity": null,
        "style": Set {},
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e" => Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f" => Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
      Object {
        "entity": null,
        "style": Set {
          "ITALIC",
        },
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;
